test/declarative_test_fwk/runner.py

```diff
@@ -0,0 +1,226 @@
+    teardown_failures = []
+    # tear down steps in reverse order
+    for step in reversed(scenario.steps()):
+        try:
+            teardown_step(step)
+        except:
+            error = f"Failed to teardown: {step}. " \
+                    f"Unexpected error: {sys.exc_info()[0]}"
+            teardown_failures.append(error)
+            logging.debug(error)
+
+    if len(teardown_failures) != 0:
+        teardown_failures.insert(0, f"Failures during teardown: {scenario}")
+        failures = "\n\t- ".join(teardown_failures)
+        logging.error(failures)
+        pytest.fail(failures)
+
+
+def run_step(step: model.Step) -> None:
+    """Runs a test scenario step
+
+    Args:
+        step: the step to run
+
+    Returns:
+        None
+    """
+
+    logging.info(f"Execute: {step}")
+    if step.verb == model.Verb.create:
+        create_resource(step)
+    elif step.verb == model.Verb.patch:
+        patch_resource(step)
+    elif step.verb == model.Verb.delete:
+        delete_resource(step)
+    wait(step)
+    assert_expectations(step)
+
+
+def create_resource(step: model.Step) -> None:
+    """Perform the Verb "create" for given test step.
+    It results in creating custom resource inside Kubernetes cluster per the specification from the step.
+
+    Args:
+        step: test step
+
+    Returns:
+        None
+    """
+
+    logging.debug(f"create: {step}")
+    if not step.input_data:
+        return
+    resource_helper = helper.get_resource_helper(step.resource_kind())
+    (reference, ko) = resource_helper.create(step.input_data, step.replacements)
+    # track created reference to teardown later
+    step.teardown_list.append((reference, ko))
+
+
+def patch_resource(step: model.Step) -> None:
+    """Perform the Verb "patch" for given test step.
+    It results in patching custom resource inside Kubernetes cluster per the specification from the step.
+
+    Args:
+        step: test step
+
+    Returns:
+        None
+    """
+
+    logging.debug(f"patch: {step}")
+    if not step.input_data:
+        return
+
+    resource_helper = helper.get_resource_helper(step.resource_kind())
+    (reference, ko) = resource_helper.patch(step.input_data, step.replacements)
+    # no need to teardown patched reference, its creator should tear it down.
+
+
+def delete_resource(step: model.Step, reference: k8s.CustomResourceReference = None) -> None:
+    """Perform the Verb "delete" for given custom resource reference in given test step.
+    It results in deleting the custom resource inside Kubernetes cluster.
+
+    Args:
+        step: test step
+        reference: custom resource reference to delete
+
+    Returns:
+        None
+    """
+
+    resource_helper = helper.get_resource_helper(step.resource_kind())
+    if not reference:
+        logging.debug(f"delete:  {step}")
+        reference = resource_helper.custom_resource_reference(step.input_data, step.replacements)
+
+    resource_helper.delete(reference)
```

Comment:
Do we need to remove the resource reference from our step's teardown_list so that double-deletes are not attempted if our test steps manually call `delete_resource`?

Reply:
As one step can only have one verb, the list is populated when verb is "create" and it is cleared during `runner.teardown_step()`